src/adm.rs

```diff
@@ -1,39 +1,269 @@
+use crate::server::ServerState;
+use crate::settings::Settings;
+use crate::tags::Tags;
+use crate::web::middleware::sentry as l_sentry;
+//use crate::server::img_storage;
+
+pub(crate) const DEFAULT: &str = "DEFAULT";
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct AdmTileResponse {
     pub tiles: Vec<AdmTile>,
 }
 
-#[derive(Debug, Deserialize, Serialize)]
+/// Filter criteria for adm Tiles
+#[derive(Default, Clone, Debug)]
+pub struct AdmFilter {
+    /// list of allowed base host strings.
+    pub filter_set: BTreeMap<String, AdmAdvertiserFilterSettings>,
+}
+
+/// The AdmAdvertiserFilterSettings contain the settings for the various
+/// ADM provided partners. These are specified as a JSON formatted hash
+/// that contains the components. A special "DEFAULT" setting provides
+/// information that may be used as a DEFAULT, or commonly appearing set
+/// of data.
+/// See `impl From<Settings>` for details of the structure.
+#[derive(Clone, Debug, Deserialize, Default)]
+pub struct AdmAdvertiserFilterSettings {
+    pub(crate) advertiser_hosts: Vec<String>,
+    pub(crate) impression_hosts: Vec<String>,
+    pub(crate) click_hosts: Vec<String>,
+    pub(crate) position: Option<u8>,
+    pub(crate) include_regions: Vec<String>,
+}
+
+pub(crate) type AdmSettings = HashMap<String, AdmAdvertiserFilterSettings>;
+
+macro_rules! check_url {
+    ($url:expr, $species:expr, $filter:expr, $tags: expr) => {{
+        let parsed: Url = match $url.parse() {
+            Ok(v) => v,
+            Err(e) => {
+                $tags.add_tag("type", $species);
+                $tags.add_extra("parse_error", &e.to_string());
+                $tags.add_extra("url", $url);
+                return Err(HandlerErrorKind::InvalidHost($species, $url.to_string()).into());
+            }
+        };
+        let host = match parsed.host() {
+            Some(v) => v.to_string(),
+            None => {
+                $tags.add_tag("type", $species);
+                $tags.add_extra("url", $url);
+                return Err(HandlerErrorKind::MissingHost($species, parsed.to_string()).into());
+            }
+        };
+        if !$filter.contains(&host) {
+            $tags.add_tag("type", $species);
+            $tags.add_extra("url", $url);
+            return Err(HandlerErrorKind::UnexpectedHost($species, host).into());
+        }
+        Ok(())
+    }};
+}
+
+impl AdmFilter {
+    /// Report the error directly to sentry
+    fn report(&self, error: &HandlerError, tags: &Tags) {
+        // dbg!(&error, &tags);
+        // TOOD: if not error.is_reportable, just add to metrics.
+        l_sentry::report(tags, sentry::event_from_error(error));
+    }
+
+    /// Check the advertiser URL
+    fn check_advertiser(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+        tags: &mut Tags,
+    ) -> HandlerResult<()> {
+        check_url!(
+            &tile.advertiser_url,
+            "Advertizer",
```

Comment:
Is this spelling with a `z` intentional?

Reply:
No, it's just that I constantly screw that up. Thanks!